Added infrastructure for perf_event_open as well as basic support for scheduler events for Linux.

florian-kuebler opened the pull request:
This PR adds basic backend support for scheduler information on Linux. It will record all events when the scheduler starts/stops executing a thread that belongs to the traced process.
On the current headless and master branch, the corresponding UI is broken, so this change will not be noticeable until this is being fixed upstream.

Internally it uses perf_event_open to record the sched:sched_switch tracepoint events, which are triggered whenever the scheduler actually switches the executed task on a concrete cpu core/hyperthread.

In order to only record events for threads that belong to the traced process, it also records all fork events, in order to keep track of the threads that belongs to this thread. For the initial perspective of the threads of the process it reads /proc/PID/tasks filesystem.

It is crucial that all events are processed in order (according to the timestamp), as otherwise, we could miss events of threads when processing their creation (fork) after the sched_switch.
As the events should be processed "live", we need to ensure, that there are always enough time between receiving an event an processing it. 
This is done by a heap (priority queue) implementation that sorts all events by timestamp, and only process the ones older than a certain OFFSET in relation the the latest event.

The main part of this PR is an abstraction layer for the perf_event_open syscall. The low-level calls to open/start/stop tracing via. perf_event_open, are done in the LinuxPerfUtils.
LinuxPerfEvent builds a class-hierarchy for all the different kind of events that will be used. This hierarchy should be extended, when switching to perf_event_open for frequency-based sampling and dynamic instrumentation (currently we have no support for callstacks).

In order to process the abstractions of perf events records, this there is a visitor pattern, that allows to  apply different processing for different events, without adding the functionality directly to the abstraction class.

pierricgimmig commented on OrbitCore/LinuxPerfEventProcessor.h:
```diff
@@ -0,0 +1,83 @@
+            // close to the max timestamp in the queue.
+            if (event->Timestamp() + OFFSET > m_MaxTimestamp) 
+                break;
+            event->accept(m_Visitor.get());
```
event->accept(visitor) ends up calling visitor->visit(event).  Could we call visitor->visit(event) directly instead of 2 virtual non inlined function calls for every event?  I'm not convinced of the added value of using a visitor pattern here.  Are you planning on having different visitor types?

florian-kuebler replied:
Due to the polymorphism, we can not directly call `visitor->visit(event)`. At the time of calling the visit/accept method, we do not know about the type of the event anymore. That's the point of the visitor patter. I choosed it, as we can indeed implement different visitors, e.g. one for sampling, one for uprobes and one for sched_events.
I mean, we could also add a virtual method "process" on the event's base class and then on each concrete class, do the interaction with the OrbitApp directly there. But then, it might be hard to keep track of state (e.g. for the callstacks of uprobes that belong to a certain function). 
I like the pattern here, as we would have a clean cut between the abstract representation of different records for perf events and processing them. Also from what I experienced so far, the two virtual calls would not hurt us that much. 

An alternative to the visitor pattern with non-virtual (or only one virtual call), would be to add an EVENT_TYPE enum value to each perf_event class and do a switch on the different Types.
But as I said, in my opinion this pattern is nice, to keep the code clean and reusable here.

pierricgimmig replied:
See https://github.com/pierricgimmig/orbitprofiler/pull/82#discussion_r348802270.

pierricgimmig commented on OrbitCore/LinuxEventTracer.cpp:
```diff
@@ -0,0 +1,126 @@
+                // perf_event_header::type contains the type of record, e.g.,
+                //  PERF_RECORD_SAMPLE, PERF_RECORD_MMAP, etc., defined in enum
+                //  perf_event_type in perf_event.h.
+                if (header.type == PERF_RECORD_SAMPLE) {
```
Should we use a switch statement instead if-else?

florian-kuebler replied:
updated the code, and use switch now

pierricgimmig commented on OrbitCore/LinuxPerfEventProcessor.h:
```diff
@@ -0,0 +1,83 @@
+    // then 1/100 sec smaller than the most recent one in the queue.
+    // This way we can ensure, that all events (from different sources)
+    // are processed in the correct order.
+    uint64_t OFFSET = 10000000 /*ns*/;
```
Where does the 1/100 sec come from, does it guarantee that events will always be processed in order?  Do you know in why we would receive events out of order?

pierricgimmig replied:
Actually, it looks like we are using a single event_buffer for different types of events.  Could it be that this is the cause of out order events?  Maybe if we had a single buffer per event type everything would always be in order and we wouldn't have to sort the events at all (and potentially get rid of the "OFFSET")?

florian-kuebler replied:
Okay let me give a short example of how the out-of-order events processing could happen.
Assume the following happens during profiling a process with PID/TID 42.
Time 0ns: TID 42 forked -> TID 13. (E1)
Time 10ns: TID 13 gets scheduled at CPU 0. (E2)

I will receive both events E1 and E2 (notice that they are of different type).
As there is no way of having events of different types (and for all cpu's) in on ring_buffer, we have the situation, that we might read E2 before we read E1. This is regardless of whether we read our buffers one after another, or all at once via. different threads. It might be simply the case that writing in the ring buffer belonging to E1 takes slightly longer then writing to the other ring buffer. Recall, that the time difference between the events might be super small. 

If we now would process E2 before E1, we would not know that TID 13 is actually a thread of our process TID 42. We would therefore, miss the scheduling event.

Thus, it is really important to put all events of all kinds into that event_buffer. Regardless of the OFFSET (I will come to this in a minute), this buffer's task is to sort all events by the timestamp and delay processing them so long, that we can sure, that we will not process a event with a higher timestamp as one that would be added slightly later. Here the visitor pattern actually helps, as we have no idea anymore, what kind of event we have in front of us, when processing them.

So finally the offset must just be large enough, such that delay processing so long, that there will be no event added (with a smaller timestamp) after we processed one with a larger timestamp.

I have to admit that this 1/100 sec. was just a hand-wavy guess of experiments and then adding a factor of 100 or 1000 ontop of what I saw. I will add code, to inform us, if that offset was sufficient and also will be fine with an offset of lets say a second. I will also rename the offset to "DELAY" at this describes it better.

pierricgimmig replied:
A lot of complexity is introduced to figure out the ProcessID from a ThreadID in order to filter the scheduling events related to our target process.  We are doing the same filtering on the Windows version, however, it is much easier since events know their PIDs.

I think we should record *all* scheduling events, not just the ones related to our process.  And that goes for Windows too.  It can help us understand why a thread is not getting scheduled on a core for example, and we'll get the full picture.

Doing this would eliminate the need of sorting events, would allow for a simpler code and would ensure that we have a *all* the scheduling information available.

Concerning the single buffer for different types of events, I don't think it's the optimal way to store the data.  I would strongly recommend cache-friendlier containers where all events of a same type would be contiguous in memory.  Under heavy load, there will likely be *a lot* of events generated and event tracking could very well become a bottleneck in the application.

Moreover, when we receive the events, we know their types.  We can easily pipe them to per-type containers where events will be tightly packed.  As it is now, using a single container of pointers to events, we lose that type information, we scatter object in memory and we need to go through virtual calls to process the data...

florian-kuebler replied:
As mentioned above, we now have system-wide and per PID tracing of scheduling events.
However, I would strongly prefer to keep that "inorder" preserving event buffer/processor. We also need this e.g. for profiling uprobes, such that we do not get out of order with uprobes and uretprobes.

pierricgimmig commented on OrbitCore/LinuxEventTracer.cpp:
```diff
@@ -0,0 +1,126 @@
+    while (!(*a_ExitRequested))
+    {
+        usleep(5000);
+
```
This comment is for the code block below btw.

florian-kuebler replied:
Yes that is correct. We do have 2 ring buffers (resulting from 2 perf_event_open calls) per cpu and the sched_switch is a high frequency one. But, we are not "processing" the events here. We just copy them over into an object and store them into the event_buffer. 
I also considered to read the data from the ring buffers from multiple threads, which would require to make the eventbuffer class be threadsafe. In particular adding simple mutex-based locks, would result in not gaining much.
Also having multiple threads (or also a smaller wait until we read again), could lead to Orbit trying to read data all the time, and thus massively increasing the performance consumption of Orbit itself (maybe affecting the performance of the profiled game).

However, the main reason why I did not go for parallel reading of the ring buffers is, that I did not noticed a single time that I started loosing events (which would be the result of on ringbuffer blocks cause of HasNewData), even in a system under stress. And my sample application actually constantly spawns new threads.

florian-kuebler replied:
Also we should not forget, that having 2*Cores+ threads also introduce a lot overhead, also when it comes to scheduling them. We might loose more than we gain here.

In general this is all a scheduling problem, right? So we could e.g. do some round robin or priority based scheduling in order to ensure, that we are "fair" and read from all buffers.
So we could solve the issue without introducing possible intrusive threading.

florian-kuebler replied:
I updated this using a form of round robbin. We now always take max 5 records from each ring buffer before continuing with the next.
In order to still be able to sleep, but still avoid having the buffers to get full, the sleep is now only executed, when there was at least one buffer that had new data.

Also experimented with that sleep. Had running the system wide tacking with load on the system (100k ContextSwitches per second -- pressed "i" on tracing) and a sleep of *0.1sec*. 
On a 10second record, only missed about 150 events in total. And still reached the sleep (i.e. did not had new event) 66 times (so out of the 6.6 seconds out of 10). So I think we should be pretty fine with sleeping 0.01 sec.

florian-kuebler commented on OrbitCore/CMakeLists.txt:
```diff
@@ -104,6 +111,11 @@ set(SOURCES
     Injection.cpp
     LinuxPerf.cpp
     LinuxPerfData.cpp
+    LinuxPerfEvent.cpp
```
Just pushed a commit to address the comments regarding the CMakeLists file and the #if __linux__ comments.

ronaldfw commented on OrbitCore/LinuxEventTracer.cpp:
```diff
@@ -0,0 +1,262 @@
+//-----------------------------------------------------------------------------
+LinuxEventTracer::LinuxEventTracer()
+{
+    //m_Callback = a_Callback ? a_Callback : [this](const std::string& a_Buffer)
```
I would either remove code that is commented out or add a comment why it is commented out and when it should be reenabled.

ronaldfw commented on OrbitCore/LinuxEventTracer.cpp:
```diff
@@ -0,0 +1,262 @@
+
+    m_Thread = std::make_shared<std::thread>(&LinuxEventTracer::foo, &m_ExitRequested);
+
+   /* m_Thread = std::make_shared<std::thread>
```
Same here.

ronaldfw commented on OrbitCore/LinuxEventTracer.cpp:
```diff
@@ -0,0 +1,262 @@
+    std::vector<std::string> tokens = Tokenize(a_Line);
+    
+    // TODO: log broken lines
+    if (tokens[4] != "sched:sched_switch:" || tokens[8] != "==>")
```
Make sure you check the length of `tokens` before accessing it. O/w this line will cause a crash.

ronaldfw commented on OrbitCore/LinuxEventTracer.cpp:
```diff
@@ -0,0 +1,262 @@
+        return;
+    }
+
+    auto oldTID = atoi(tokens[1].c_str());
```
Same here and below. Make sure that `tokens` is actually long enough.

ronaldfw commented on OrbitCore/LinuxEventTracer.cpp:
```diff
@@ -0,0 +1,262 @@
+{
+    std::vector<std::string> tokens = Tokenize(a_Line);
+    
+    // TODO: log broken lines
```
Do we track TODOs as open issues?

florian-kuebler replied:
Actually, not yet (at least for all of them). I will try to keep them in-sync for the larger ones and address small ones that are left immediately.

ronaldfw commented on OrbitCore/LinuxEventTracer.cpp:
```diff
@@ -0,0 +1,262 @@
+        CS.m_Time = time;
+        CS.m_ProcessorIndex = cpu;
+        //TODO: Is this correct?
+        CS.m_ProcessorNumber = cpu;
```
What's the meaning of ProcessorIndex vs. ProcessorNumber? Does it make sense to track them individually? (Are there places where they are actually different?)

florian-kuebler replied:
ProcessorNumber is the number of CPUs in the machine, ProcessorIndex is the core index. Currently ProcessorNumber is not used by Orbit at all. In particular, I don't think that this information is available while tracing. We could use 0 for ProcessorNumber otherwise. In the end, it basically needs to be consistent.

ronaldfw commented on OrbitCore/LinuxEventTracer.cpp:
```diff
@@ -0,0 +1,262 @@
+    }
+}
+
+void LinuxEventTracer::foo(bool* a_ExitRequested)
```
Name this properly.

ronaldfw commented on OrbitCore/LinuxEventTracer.cpp:
```diff
@@ -0,0 +1,262 @@
+        fds.push_back(LinuxPerfUtils::task_event_open(cpu)); 
+        fds.push_back(LinuxPerfUtils::tracepoint_event_open(tracepointID, -1 /*pid*/, cpu)); 
+    }
+
```
Remove one of the empty lines here? Not sure what the code style here should be.

ronaldfw commented on OrbitCore/LinuxEventTracer.cpp:
```diff
@@ -0,0 +1,262 @@
+
+    // TODO: wrap the ringbuffers into a class
+    std::vector<perf_event_mmap_page*> metadatas;
+    std::vector<char*> ring_buffers;
```
Naming of variables mixes camelCase and underscore_naming.

florian-kuebler replied:
should be resolved with the latest commit

ronaldfw commented on OrbitCore/LinuxEventTracer.cpp:
```diff
@@ -0,0 +1,262 @@
+        ring_buffers.push_back(buffer);
+        uint64_t buffer_length = metadata->data_size;
+        ring_buffer_lengths.push_back(buffer_length);
+
```
Remove one empty line?

ronaldfw commented on OrbitCore/LinuxEventTracer.cpp:
```diff
@@ -0,0 +1,262 @@
+        targetProcess->AddThreadId(tid);
+    }
+
+    // create a priority queue to buffer events an process them in sorted by their timestamp
```
2 typos: "...and process them sorted by their timestamp"

ronaldfw commented on OrbitCore/LinuxEventTracer.cpp:
```diff
@@ -0,0 +1,136 @@
+
+    while (!(*a_ExitRequested))
+    {
+        usleep(5000);
```
What's the reason for the sleep?

florian-kuebler replied:
Without the sleep, this thread would constantly read the metadata from the buffer in order to check if there are new events. Reading the header would be done in a much higher frequency than new events can even occur, and we would wast a lot of CPU time, basically useless.

ronaldfw replied:
I would add a comment that explains this in the code.

ronaldfw commented on OrbitCore/LinuxEventTracer.h:
```diff
@@ -0,0 +1,26 @@
+    void CommandCallback(std::string a_Line);
+
+private:
+    bool m_ExitRequested = true;
```
Seems counterintuitive that this is set to `true` by default, that is, "exit has been requested". Maybe the name is not chosen well? The meaning of this variable seems to be, the tracer is running.

florian-kuebler replied:
Well, I agree that the default value with true, is not intuitive.
However, the main usage of that variable is in the while(!m_ExitRequested) that basically runs until the User presses "X" in the CaptureView in order to stop tracing. In particular, we would not stop until all the remaining events has been processed. So it is more about the Request from the user, as about the current state (i.e. whether it is running or not).

Basically, I copied the name from the Bpftrace implementation. However, I am happy to change it, to something more meaningful.

ronaldfw commented on OrbitCore/LinuxEventTracerVisitor.cpp:
```diff
@@ -0,0 +1,50 @@
+        CS.m_Time = e->Timestamp();
+        CS.m_ProcessorIndex = e->CPU();
+        //TODO: Is this correct?
+        CS.m_ProcessorNumber = e->CPU();
```
I commented already elsewhere, but not sure if that comment is still visible: Is there a reason we have both ProcessorIndex and ProcessorNumber? Where is this actually different?

florian-kuebler replied:
Yes it is still visible, but as this is the current version, I copy over my response and close the one on the outdated commit:

ProcessorNumber is the number of CPUs in the machine, ProcessorIndex is the core index. Currently ProcessorNumber is not used by Orbit at all. In particular, I don't think that this information is available while tracing. We could use 0 for ProcessorNumber otherwise. In the end, it basically needs to be consistent.

ronaldfw commented on OrbitCore/LinuxPerfEvent.h:
```diff
@@ -0,0 +1,132 @@
+
+#include "PrintVar.h"
+#include "LinuxPerfUtils.h"
+
```
I would remove this empty line.

ronaldfw commented on OrbitCore/LinuxPerfEventProcessor.h:
```diff
@@ -0,0 +1,79 @@
+
+// A comparator used for the priority queue, such that pop/top will
+// always return the oldest event in the queue.
+class TimeStampCompare
```
"Timestamp" (as in other places), not "TimeStamp"

ronaldfw commented on OrbitCore/LinuxPerfEventProcessor.h:
```diff
@@ -0,0 +1,79 @@
+{
+public:
+    // While processing, we do not touch the events with a timestamp less 
+    // then 1/100 sec smaller than the most recent one in the queue.
```
```suggestion
    // than 1/100 sec smaller than the most recent one in the queue.
```

ronaldfw commented on OrbitCore/LinuxPerfEventProcessor.h:
```diff
@@ -0,0 +1,79 @@
+    // then 1/100 sec smaller than the most recent one in the queue.
+    // This way we can ensure, that all events (from different sources)
+    // are processed in the correct order.
+    uint64_t OFFSET = 10000000 /*ns*/;
```
Consider adding the unit "ns" to the OFFSET name, like OFFSET_IN_NS, or something like that.

ronaldfw commented on OrbitCore/LinuxPerfEventVisitor.h:
```diff
@@ -0,0 +1,13 @@
+#include "LinuxPerfEvent.h"
+
+// TODO: As the class hierachy is not fixed yet, we need to 
+//  keep this in sync.
```
In sync with what?

ronaldfw commented on OrbitCore/LinuxPerfRingBuffer.cpp:
```diff
@@ -0,0 +1,65 @@
+    assert(header->type != 0);
+    assert(m_Metadata->data_tail + header->size <= m_Metadata->data_head);
+}
+
```
Remove one empty line?

ronaldfw commented on OrbitCore/LinuxPerfRingBuffer.cpp:
```diff
@@ -0,0 +1,65 @@
+    if (count > m_BufferLength) {
+        // If mmap has been called with PROT_WRITE and
+        // perf_event_mmap_page::data_tail is used properly, this should not happen,
+        // as the kernel would no overwrite unread data.
```
"not" or "now"? ("no" seems like a typo)

ronaldfw commented on OrbitCore/LinuxUtils.cpp:
```diff
@@ -46,6 +46,7 @@ namespace LinuxUtils {
 std::vector<std::string> ListModules( uint32_t a_PID )
 {
     std::vector<std::string> modules;
+    // TODO: we should read the file directly instead
```
Or memory map the file?

ronaldfw commented on OrbitCore/LinuxPerfUtils.h:
```diff
@@ -0,0 +1,99 @@
+
+
+// This struct must be in sync with the SAMPLE_TYPE_FLAGS.
+// TODO: rename T
```
Is this done?

ronaldfw commented on OrbitCore/LinuxPerfUtils.h:
```diff
@@ -0,0 +1,99 @@
+namespace LinuxPerfUtils
+{
+    // TODO: these constants should be made available in some settings
+    static const uint32_t STACK_SIZE = (1u << 16u) - 8; // TODO: the stack is not yet recorded
```
Does the TODO mean that the stack size here is incorrect until the stack has been recorded? Maybe make this is a constant whose meaning is actually always correct and then record the actual stack size in a variable that always contains the correct value.

florian-kuebler replied:
The stack size is actually correct, but it would simply not used in perf_event_open internally.

ronaldfw commented on OrbitCore/LinuxPerfUtils.cpp:
```diff
@@ -0,0 +1,81 @@
+{
+    const size_t PAGE_SIZE = getpagesize();  // 4 KB
+
+    // "The mmap size should be 1+2^n pages, where the first page is a meta‐
```
What's the source for this comment in quotes?

florian-kuebler commented on OrbitCore/LinuxEventTracer.cpp:
```diff
@@ -0,0 +1,126 @@
+                    //  i.e. for multiple perf_event_open calls, we need be able to
+                    //  distinguish the different ring buffers.
+                    auto sched_switch = ring_buffer->ConsumeRecord<perf_event_record<sched_switch_tp>>(header);
+                    auto e = std::make_shared<LinuxSchedSwitchEvent>(sched_switch);
```
I really wanted to avoid direct memory management here, as I don't think it is needed.
But sure, I have to admit, that especially the shared_ptr comes with some overhead.
Would it be an option for you, to go for unique_ptrs instead of shared_ptrs? They are way more lightweight. This would only require some tweaking of the EventProcessor datastructure, that has a priority queue internally.

As we really need to be careful, to sort the events from all the different ring_buffers and process them in sync with their timestamps. So I can't just add them into a Blockchain class.
Maybe, I could add them there and only store the pointers in the queue, but then the blocks would get fragmented, as the events are not necessarily processed (and therefore "freeed" afterwards) in the order they were  allocated. So they might not been removed from the blockchain in the (reverse-) order as they were allocated.

pierricgimmig replied:
See https://github.com/pierricgimmig/orbitprofiler/pull/82#discussion_r348802270.

florian-kuebler replied:
Having a system-wide mode now, that directly converts the events into ContextSwitches and send them to the backend, where they were stored cache friendly.
Despite that, I use unique_ptrs instead of shared_ptrs in the event processor now. This should be way better in terms of overhead. Still not cache friendly though. However, did not noticed any issue so far.

florian-kuebler commented:
> There's one comment open where I would leave a comment in the code about the sleep in the loop (why is this done? possibly: what are reasonable values for the length of the sleep?). Apart from that looks good to me (obviously I don't know the code as well as Pierric, so he'll have to approve).

Thanks again for the review. I did some more experiments and adjusted the the sleep to by 1000 microseconds instead. Also added a comment in the code.

pierricgimmig commented:
Hi Florian, if you sync latest you should be able to see the context switch events at the top of the capture window.

danielfenner commented on OrbitCore/LinuxPerfRingBuffer.cpp:
```diff
@@ -0,0 +1,86 @@
+
+    // also we can optimize index / m_BufferLength to:
+    uint64_t index_div_length = 
+        (index + ((index >> 31) & ((1 << exponent) + ~0))) >> exponent;
```
"31" also looks fishy

danielfenner commented on OrbitCore/LinuxPerfUtils.h:
```diff
@@ -0,0 +1,124 @@
+
+    // This struct must be in sync with the SAMPLE_TYPE_FLAGS.
+    template<typename raw_data_t, typename registers_t, typename stack_t>
+    struct __attribute__((__packed__)) perf_event_record {
```
Can we somehow implement the "accept" method here such that we don't need to copy twice?

florian-kuebler replied:
I just pushed a solution, where I use the LinuxPerfEvent (sub)classes directly as a target for the memcpy from the ringbuffer.

florian-kuebler commented on OrbitCore/LinuxEventTracer.cpp:
```diff
@@ -0,0 +1,193 @@
+                            // we need to know the type of the perf_record_sample at compile time,
+                            //  i.e. for multiple perf_event_open calls, we need be able to
+                            //  distinguish the different ring buffers.
+                            event_buffer.Push(std::make_unique<LinuxSchedSwitchEvent>(sched_switch));
```
It might be cheaper to copy the pointers in the priority_queue (when inserting a new element and moving the other elements around) instead of copying the complete objects. On the other side, the elements would be then all in the same range of memory (due to the underlying vector), which is more cache friendly.

We should do micro benchmarks here. For the moment, I'll stick with the pointers and add a comment here.

florian-kuebler commented on OrbitCore/LinuxPerfUtils.cpp:
```diff
@@ -0,0 +1,93 @@
+    // http://man7.org/linux/man-pages/man2/mmap.2.html
+    // Use mmap to get access to the ring buffer.
+    void* mmap_ret =
+        mmap(nullptr, mmap_length, PROT_READ | PROT_WRITE, MAP_SHARED, a_FileDescriptor, 0);
```
I added it in the destructor of the ringbuffer class

florian-kuebler commented on OrbitCore/LinuxPerfRingBuffer.h:
```diff
@@ -0,0 +1,40 @@
+class LinuxPerfRingBuffer
+{
+public:
+    LinuxPerfRingBuffer(uint32_t a_PerfFileDescriptor);
```
I needed to make this class a move only class. Could you have a look at the change?

danielfenner replied:
Move only sounds good: You want the mmap being owned exclusively.

florian-kuebler commented on OrbitCore/LinuxEventTracer.cpp:
```diff
@@ -0,0 +1,193 @@
+    // stop capturing
+    for (int32_t fd : fds)
+    {
+        LinuxPerfUtils::stop_capturing(fd);
```
I decided to put it in the destructor of the buffer.

danielfenner commented on OrbitCore/LinuxPerfRingBuffer.cpp:
```diff
@@ -0,0 +1,145 @@
+//-----------------------------------------------------------------------------
+LinuxPerfRingBuffer& LinuxPerfRingBuffer::operator=(LinuxPerfRingBuffer&& other) noexcept
+{
+    std::swap(m_Metadata, other.m_Metadata);
```
The swap is a bit of a surprise. Why not move?

florian-kuebler replied:
We do not move, as we need to keep the other buffer intact, until the destructor calls munmap.

danielfenner commented on OrbitCore/LinuxPerfRingBuffer.cpp:
```diff
@@ -0,0 +1,145 @@
+    const uint64_t index_count = index + a_Count - 1;
+    // and also (index + a_Count - 1) / m_BufferLength
+    const uint64_t index_count_div_length = 
+        (index_count + ((index_count >> 31) & ((1 << exponent) + ~0))) >> exponent;
```
"31" and "~0" are inconsistent with the otherwise identical line above.

florian-kuebler commented on OrbitCore/LinuxPerfEventProcessor.cpp:
```diff
@@ -0,0 +1,48 @@
+
+        // We should not read all events, otherwise we could miss events
+        // close to the max timestamp in the queue.
+        if (event->Timestamp() + DELAY_IN_NS > m_MaxTimestamp) 
```
I added comments and TODOs in the code
